File: DukeMapGenLib/src/java/trn/duke/experiments/stonetunnels/ExitBlock.java
package trn.duke.experiments.stonetunnels;

import org.apache.commons.lang3.tuple.Pair;

import trn.*;
import trn.duke.TextureList;
import trn.duke.experiments.gridblock.AbstractBlock;
import trn.duke.experiments.gridblock.Block;
import trn.duke.experiments.gridblock.Connector;
import trn.duke.experiments.gridblock.NorthSouthConnector;
import trn.maze.Heading;

/**
 * 
 * @author Dave
 *
 */
public class ExitBlock extends AbstractBlock implements Block {

	// TODO apparently any nonzero value will work, and buildhlp claims it should be 32767....
	public static final int NUKE_BUTTON_END_LEVEL = 65535;

	private Connector connector;
	
	private final Heading connectorEdge = Heading.NORTH;
	
	private final int floorZ = StoneConstants.UPPER_FLOORZ;

	public ExitBlock(Pair<Integer, Integer> gridCoordinate) {
		super(gridCoordinate);

		this.connector = NorthSouthConnector.northEdge(this);
	}

	@Override
	public Connector getConnector(Heading heading) {
		return heading == connectorEdge ? connector : null;
	}

	@Override
	public int draw(Map map) {
		
		int south = getSouthEdge();
		int west = getWestEdge();
		int north = getNorthEdge();
		int east = getEastEdge();
		
		PointXY[] box = new PointXY[]{
				new PointXY(west, south),
				new PointXY(west, north),
				new PointXY(east, north),
				new PointXY(east, south)
			};
		
		int sectorIndex = map.createSectorFromLoop(Wall.createLoop(box, StoneConstants.UPPER_WALL));
		
		StoneConstants.UPPER_SECTOR.writeTo(map.getSector(sectorIndex));
		
		
		Sprite exitSprite = new Sprite(
				(east + west) / 2,
				south,
				256 << 4, //z
				(short)sectorIndex);
		
		exitSprite.setTexture(TextureList.Switches.NUKE_BUTTON);
		exitSprite.setCstat((short)Sprite.CSTAT_FLAGS.PLACED_ON_WALL);
		exitSprite.setXRepeat(21);
		exitSprite.setYRepeat(26);
		exitSprite.setAngle(AngleUtil.ANGLE_UP);
		exitSprite.setLotag(NUKE_BUTTON_END_LEVEL);
		
		map.addSprite(exitSprite);
		
		
		connector.setSectorIndex(sectorIndex);
		
		
		
		return -1;
	}

}

File: DukeMapGenLib/src/scala/duchy/experiments/render/maze/stonetunnels/ExitBlock.scala
package duchy.experiments.render.maze.stonetunnels

import org.apache.commons.lang3.tuple.Pair
import trn.duke.TextureList
import trn.{BuildConstants, PointXY, Wall, Sprite, AngleUtil}
import trn.duke.experiments.gridblock.{LegacyConnector, AbstractBlock}
import trn.maze.Heading

class ExitBlock(gridCoordinate: Pair[Integer, Integer]) extends AbstractBlock(gridCoordinate) {

  val connector = LegacyConnector.northEdge(this)

  val connectorEdge = Heading.NORTH;  // rotation didn't exist when I originally wrote this
  val floorZ = StoneConstants.UPPER_FLOORZ;

  override def getConnector(heading: Heading): LegacyConnector = if (heading == connectorEdge) { connector } else { None.orNull }

  /**
    *
    * @param map
    * @return the index of the (a?) sector that was created.
    */
  override def draw(map: trn.Map): Int = {
    val south = getSouthEdge
    val west = getWestEdge
    val north = getNorthEdge
    val east = getEastEdge
    val box = Array[PointXY](new PointXY(west, south), new PointXY(west, north), new PointXY(east, north), new PointXY(east, south))
    val sectorIndex = map.createSectorFromLoop2(Wall.createLoopAsList(box, StoneConstants.UPPER_WALL))
    StoneConstants.UPPER_SECTOR.writeTo(map.getSector(sectorIndex))

    val exitSprite = new Sprite((east + west) / 2, south, 256 << 4, //z
      sectorIndex.toShort)

    exitSprite.setTexture(TextureList.Switches.NUKE_BUTTON)
    exitSprite.setCstat(Sprite.CSTAT_FLAGS.PLACED_ON_WALL.toShort)
    exitSprite.setXRepeat(21)
    exitSprite.setYRepeat(26)
    exitSprite.setAngle(AngleUtil.ANGLE_UP)
    exitSprite.setLotag(TextureList.Switches.NUKE_BUTTON_LOTAG) // how can 65535 be correct if its a short?

    map.addSprite(exitSprite)

    connector.setSectorIndex(sectorIndex)

    return -1
  }
}
